### ISSUE TITLE
get_fval returns inf and get_grad returns all NaNs

### ISSUE THREAD

**Issue by thomassligon** [ICB-DCM/pyPESTO#659]
In site-packages/pypesto/objective/base.py
get_fval at line 261
get_grad at line 268
Input: x of length 572 looks normal, but
get_fval returns inf, and
get_grad returns all NaNs.
The next call to amici fails because of the NaNs.

What should these routines do?

at get_fval, x = [0:572]
...
returns fval = inf

at get_grad, x = 
array([-6.26449451e+00, -5.29009814e+00, -4.56026855e+00,  3.10531181e+00,
        2.52331119e+00,  3.60818385e+00,  3.31940002e+00,  1.69150026e+00,
        3.28965034e+00, -1.00332056e+00, -5.94530586e+00, -9.69175360e-01,
       -5.50081273e+00, -3.78303155e+00, -6.92050678e+00, -2.10071047e+00,
        3.64684272e-01, -2.12291528e+00, -4.31246625e-01, -2.89563387e+00,
       -1.65414253e+00, -5.93491223e+00,  3.92002968e-01, -1.40020601e-02,
       -2.14769208e+00, -1.94427508e+00, -1.89489566e+00, -6.84728821e+00,
       -1.71254521e+00, -4.66427998e+00, -4.11783230e+00, -1.00034956e-01,
       -1.33585456e+00, -3.11721462e+00, -2.01368077e+00, -4.74062764e+00,
       -6.03909340e+00, -2.11722078e+00, -3.08765322e+00, -3.42102765e+00,
       -5.14682474e+00, -4.69430303e+00, -2.52028249e+00, -2.65674102e+00,
       -3.40165987e+00, -4.42921470e-01, -3.58586760e+00, -6.63197066e+00,
       -1.11403749e+00, -1.11083525e-01, -3.61535546e-01, -1.02073082e+00,
       -5.00539540e+00, -5.20004726e+00, -2.72645806e+00, -3.82275149e+00,
       -4.33572704e+00, -5.56613707e+00, -2.41225917e+00, -2.92199622e+00,
       -3.01043455e+00, -5.43668871e+00, -1.92460013e+00, -1.62572019e+00,
       -1.98900460e+00, -2.69999511e+00, -2.97353883e+00, -3.62941853e+00,
       -1.37214634e-01, -1.39191470e+00, -6.72352098e+00, -6.36803859e+00,
       -4.54422254e+00, -2.74680770e+00, -3.87962538e+00, -3.42615761e+00,
       -4.22114750e+00, -4.95196546e+00, -3.88143149e+00, -3.29315371e+00,
       -1.51822882e-01, -1.79667074e+00,  1.11948166e+00, -6.26653676e+00,
        5.29662192e-01, -1.06627766e+01,  3.60728495e-01, -1.33573525e+00,
        2.31801717e-01, -3.79883086e+00, -2.22801170e+00, -5.60138201e+00,
        9.71239234e-01, -5.93939104e+00,  1.86700087e+00, -4.96853135e+00,
       -5.64854259e+00, -3.52057204e+00, -8.09326175e-01, -2.50211298e+00,
       -6.19754889e+00,  1.34433766e+00, -5.85506218e+00,  1.26938173e+00,
       -7.81223269e+00, -8.11048542e-01, -6.66139499e+00,  1.94516318e+00,
       -6.60513625e+00,  1.97125607e+00, -4.79710922e+00,  8.50432822e-01,
       -8.51613338e+00,  1.86699328e-01, -3.05109969e+00,  6.48354378e-01,
       -2.63606129e+00, -4.09167819e+00, -4.87405229e+00, -2.82752108e+00,
       -4.39637763e+00, -5.71101243e+00, -4.93405595e+00,  8.13256213e-01,
       -1.71549532e+00, -1.22277748e+00, -3.46375873e+00, -7.39947340e-01,
       -6.48504217e+00,  2.76910770e+00, -3.13944883e-01, -8.99114735e-01,
       -4.76652889e+00, -1.18396614e+00, -6.05144273e+00, -5.79901479e+00,
       -4.20629717e+00, -4.61996883e+00, -5.39303995e-01, -3.76492028e+00,
       -1.69029740e+00, -4.32379985e+00, -8.24104379e-01, -6.51669574e+00,
       -3.19312547e+00, -5.37464713e+00, -3.47591638e-03, -1.21739937e+00,
       -6.96191400e+00, -4.59683758e+00, -9.50005040e-01, -1.80737287e+00,
        1.48438102e+00, -4.29409454e+00, -4.05350642e+00, -1.61007913e+00,
       -5.19584188e+00, -1.30261536e+00, -1.78699387e+00, -2.77956437e+00,
       -2.50068162e+00, -2.60183338e+00, -8.37322978e-01,  5.88608053e-01,
       -4.63272986e+00,  1.41874720e+00, -5.32533073e+00, -6.54595543e-01,
       -3.83303832e+00, -6.28976160e+00, -3.64692001e+00, -2.71396661e-01,
       -8.54141041e-01, -2.15259492e+00,  1.97782664e+00, -4.59406891e+00,
        3.55960252e-01, -3.31778977e+00, -6.66903358e+00, -8.62673711e-01,
       -2.07423922e+00,  5.95842390e-03, -5.00322009e+00, -2.82750444e+00,
       -2.73578581e+00, -6.45213946e+00, -2.98403632e-01, -3.07862717e+00,
       -1.00541120e+00, -1.59537835e+00, -1.88756527e+00, -4.46489103e+00,
       -2.34271219e-01, -3.34159096e-01, -3.88447578e+00, -5.49933352e+00,
       -6.12783311e-01, -2.56617056e+00, -3.99726798e+00, -1.20347238e+00,
       -5.59215867e+00, -6.44825176e+00, -2.36266378e+00, -6.16535506e+00,
       -3.20655947e+00, -2.83678465e+00, -4.37175216e+00, -6.66549653e+00,
       -1.26593036e+00, -6.44535322e+00, -3.70275941e+00, -3.06655931e+00,
       -6.32679914e-01, -3.69088480e+00, -4.34382647e+00, -1.50673314e+00,
        8.52607421e-01, -1.51715196e+00, -2.55875424e+00, -3.28711550e+00,
       -2.40235510e+00, -2.35560602e+00, -1.03324007e+00, -3.51647825e+00,
       -1.89064087e+00, -3.52781385e+00, -6.85502616e+00, -3.27314540e+00,
       -4.77386241e+00, -6.03331492e-01, -6.80206891e+00, -3.62648260e+00,
       -3.03040682e+00, -4.53628025e+00, -4.82192976e+00, -8.38812870e+00,
       -3.03967636e+00, -3.66224865e+00, -6.96008432e+00, -3.42402253e+00,
       -1.47767123e+00, -2.76399598e+00, -6.45931859e+00, -3.10861554e+00,
       -8.19906753e-01, -2.74908304e+00, -5.95711449e+00, -4.51200315e+00,
       -1.93401932e+00,  9.09977579e-01, -1.71489832e+00, -2.70728092e+00,
       -2.92943306e+00,  7.08257498e-02, -3.80548195e+00, -3.52707601e+00,
       -3.95347171e+00, -6.62570344e+00, -3.52817864e+00, -1.27775176e+00,
       -1.83848560e+00, -6.04950637e+00, -1.51814634e+00, -6.02908362e+00,
       -1.95000060e+00, -3.13059603e+00, -2.09442056e+00, -6.71041910e+00,
        5.90014923e-01, -5.84571110e+00, -6.75406461e+00, -5.22654108e+00,
       -1.03768903e+00, -6.04889809e+00, -1.04453438e+00, -3.12496860e+00,
       -2.87601009e+00, -4.38748635e+00, -2.87571683e-03, -2.09782525e+00,
       -2.64633245e+00, -2.87339125e+00, -1.61113464e+00, -1.45988092e+00,
       -1.46948909e-01, -3.63665478e+00, -3.65744634e+00, -2.54731480e+00,
       -1.92340006e+00,  1.83229811e-01, -2.89752151e+00, -1.10462853e+00,
       -2.56647792e+00, -2.12428556e-01, -2.98821013e+00, -5.88435619e+00,
       -3.11249442e+00, -1.05610009e+00, -4.92554301e-01, -2.15620877e+00,
       -3.26839083e+00, -1.16661036e+00, -2.44749612e+00, -2.98182113e+00,
       -9.14219536e-01, -1.36028141e+00, -3.01489941e+00,  3.92684539e-01,
       -4.22177255e-02, -3.31990276e+00, -3.29848129e+00, -5.69686626e+00,
       -6.76352213e+00, -3.41547405e-01, -6.70498890e+00, -2.83629391e+00,
       -6.91345298e+00, -5.59047942e+00, -6.62705215e+00, -1.82224751e+00,
       -3.24932979e+00, -1.68903558e+00, -3.52285405e+00, -4.05332805e+00,
       -6.75007068e+00, -6.84413948e-01, -6.04734606e+00, -5.93452408e+00,
       -3.92707161e+00, -3.04156643e+00, -5.80527368e+00, -5.04750024e+00,
       -4.92155209e+00, -4.39181759e+00, -5.92959884e+00, -5.19389664e+00,
       -5.34757194e+00, -1.87466223e+00, -3.63175082e+00, -1.46342551e+00,
       -4.52030576e+00, -2.97976447e+00, -5.96725857e+00, -5.47790679e+00,
       -1.11777989e+00, -2.68222302e+00, -2.25639348e+00,  9.97716577e-01,
       -1.74065075e+00, -1.85485756e+00, -2.38078657e+00,  4.30282321e-01,
       -2.42040975e+00, -6.94099526e-01, -2.81977398e-01,  6.52913847e-01,
       -3.05461539e+00, -5.49118487e+00,  5.55562569e-01, -9.88293503e-01,
       -6.41113720e-01, -1.73074801e+00, -2.09733286e+00, -2.48938700e+00,
        6.04697770e-01, -3.14333484e+00, -3.03723924e+00, -6.79961315e+00,
       -2.67240536e+00, -3.51792104e+00, -3.80091985e+00, -3.25983111e+00,
       -1.88474751e+00, -3.11150376e+00, -1.43714646e+00, -4.82752669e+00,
        2.04518896e-01, -1.67809777e+00, -5.90518043e+00, -7.14893183e-01,
       -2.80407874e+00, -3.05947131e+00, -6.56993442e-01, -8.80047399e-01,
       -1.34794528e+00, -2.15051804e+00,  4.30225869e-01, -1.69456685e+00,
        3.07751334e-01, -8.43500515e-01, -3.15861172e+00, -6.20801021e+00,
       -9.39291760e-01, -6.39534338e-01, -5.39314055e+00,  2.39694757e+00,
       -5.55190891e+00, -2.57547487e+00, -4.66912066e+00,  1.58900796e+00,
        2.38056959e+00,  3.13239924e+00,  1.24692891e+00,  1.73168897e+00,
        2.08665175e+00,  1.40068438e+00, -2.47271460e-02,  7.61139722e-01,
        9.61393202e-01, -7.96006786e-01, -1.07313524e+00, -7.38597701e-01,
       -4.36413841e+00, -1.91519518e+00, -5.42417104e+00, -2.98139100e+00,
        2.64933723e+00,  2.59606579e+00,  4.64237845e+00,  9.23115035e-01,
       -1.07711241e+01,  8.40726544e-02, -3.45768698e-01, -9.00672550e-01,
       -3.05029605e+00, -3.46885021e-02,  1.26281612e+00,  1.81534448e+00,
        2.35576629e-01, -7.33753499e-01,  2.82758180e+00,  5.97118470e-02,
        2.62356742e+00,  5.57948285e-01,  2.06712087e+00, -1.14926768e+00,
       -9.58131668e-01,  5.87502979e-01,  2.97889842e+00,  2.61562869e+00,
       -1.05382936e+00,  2.14314124e+00,  7.26816318e-01,  1.54167918e+00,
       -1.32067572e-02,  2.58231349e-01, -8.56018251e-01, -4.36469261e+00,
       -3.09691196e+00, -9.71628399e-01,  1.93864505e+00,  2.97068365e+00,
       -3.40030779e-01,  1.38843529e+00, -2.52251617e-01, -4.56031601e-01,
        4.94063378e-01, -1.43625425e-01, -5.19057799e-01, -6.24576376e+00,
       -5.73580294e+00, -6.76262600e+00, -3.15739805e+00,  2.98984671e+00,
        7.23385428e-01,  2.43600760e+00,  3.57405264e+00,  2.37754258e+00,
        1.16379898e+00,  2.43540893e+00, -4.84252534e+00, -1.05618694e+00,
        4.18777143e-01, -9.73050342e-01, -6.79093593e-01, -2.74659789e+00,
       -4.05320958e+00, -3.39171711e-01, -3.07960982e+00,  8.40493724e-01,
        1.21870101e+00, -5.05105859e-01, -6.18775301e+00,  2.54542062e+00,
        4.17574864e-01,  1.98609639e+00, -1.50097151e+00, -7.81888115e+00,
       -2.23060002e-02, -3.61424144e-01,  6.24559672e-01, -6.03603809e+00,
        5.74922543e-01,  3.30570080e+00,  2.38014854e+00,  2.96503663e+00,
       -1.22170448e+00, -9.17225619e+00,  1.98737362e+00, -9.71968057e+00,
        1.16332543e+00,  2.18101456e+00,  8.69156228e-02, -5.80611056e+00,
        1.20071807e+00, -4.32671123e-01, -6.16757211e+00,  2.96722963e-01,
       -5.62186067e-01, -7.49048208e+00,  4.65971521e-01,  3.97200356e+00,
        3.33611466e-01, -6.50447719e+00, -3.17499161e+00, -4.61302567e+00,
       -3.00372680e+00, -1.27435599e-01, -6.61061464e+00,  4.87242557e-01,
       -5.96163461e+00, -3.10950426e+00, -3.93393628e+00, -3.63656985e+00,
       -4.97396094e+00, -3.22028154e+00,  2.57963586e-01, -5.68730784e+00,
       -2.00682202e+00, -1.17664046e+00,  3.28150916e+00, -6.46009420e-01,
        9.06377728e-01,  8.40100744e-01,  2.43091109e+00, -3.42041358e-01,
        7.80995997e-01, -1.58751475e+00, -6.31385697e+00, -3.34757792e+00,
       -2.63909509e+00, -2.78216091e+00, -5.48116065e+00, -6.72077711e+00,
       -6.27827853e+00, -3.51442859e+00, -1.89863331e-01, -3.89555157e+00,
       -2.72093242e-01, -3.90909786e+00, -2.71698344e+00, -2.76625103e+00,
       -5.39510104e-01, -3.17495323e+00, -1.20640755e+00, -9.58111924e+00,
        1.17578982e+00, -1.69165904e+01, -6.43461292e+00, -6.31583713e+00,
       -9.13589161e-01, -6.67571914e+00,  7.25283236e-01, -3.53417828e+00,
        1.64277163e+00, -6.69858814e+00,  1.97850130e-01, -4.62900522e+00,
        1.19749677e+00, -4.35692326e+00, -1.02361765e+00,  3.68783728e+00])
returns grad = 
array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan])

**Comment by jvanhoefer** [comment]
Hey Tom, 
your line-numbering does not align with the content in the current master branch (that is shifted there), suggesting that you are not using the current version of pypesto, so `pip3 install --upgrade pypesto` or `pip3 install --upgrade amici` might be a first step. 

Furthermore: It seems as if the objective function is not able to perform the forward simulation. This could be due to a "problematic" model, that e.g. explodes for some parameter values. Hence

* trying more start points
* Looking at the specific value, at which the ODE solution broke (given  in the error message, that is printed to your command line, when an optimization run fails...) can give you a hint, what is going wrong. Also simulating the model with AMICI for parameters, where the gradient evaluation failed might give you a hint which species etc. is exploding.

All in all, this is often more a problem of the model/simulation and not particularly of the fitting itself...

**Comment by thomassligon** [comment]
Hi Yannik, Jakob,

thanks for the tips! I will try parameter scan in Copasi and see if that shows me anything. But, I was hoping to get more information around the calls to get_fval and get_grad. If this were giving me a few NaNs, that would tell me something about which reactions are causing problems, but all NaNs doesn't.
What I have done up to now: In Copasi, steady state succeeds, but only after it tries Newton and forward simulation multiple times. After calling PETab/PyPESTO minimize, amici tries to run steady-state, produces two "normal" integration errors, then calls get_fval and get_grad for the first time, and both of them fail completely. I found this spot by testing and looking at the call stack in the python debugger, since I do not know the logic PyPESTO is using. I am surprised that the line number doesn't match, because I updated all python packages before running the last test. I will check this again.
In Copasi (and Cell Designer), I have used biologically plausible values for rates of proteins based on transcription and translation rates and, for reactions involving other substances, such as phosphatidylcholine or 15-oxo-ETE, I have adjusted the parameters to make sure that the concentration and reactions are non-zero. After running steady-state in Copasi, I changed the initial concentrations of species to match the steady-state value. Forward simulation is working without any problems, and the curves are very flat, which looks like a steady state. 
This is also issue #1371 in AMICI.

**Comment by yannikschaelte** [comment]
Hi, get_fval, get_grad are just very simple wrappers around https://github.com/ICB-DCM/pyPESTO/blob/main/pypesto/objective/amici.py#L306.
So AMICI simulation for the same parameter values succeeds, but when used in pyPESTO it fails? That would indicate that something about the assignment of data or parameters could have gone wrong. You can get the complete amici output (RData) from the objective by calling `objective(..., return_dict=True)`, see https://github.com/ICB-DCM/pyPESTO/blob/main/pypesto/objective/base.py#L98. Maybe this helps to see whether e.g. the states `x`, observables `y` and residuals `res` have finite values?

**Comment by thomassligon** [comment]
Thanks for all the help. I considered a parameter scan in Copasi, but that requires specifying which parameter to scan, and there are 572 parameters, so scanning each one individually would probably take some time.
Instead, I tried restricting the range of parameters used by optimize. Currently, I am using the initial value of the parameter in SBML and specifying the upper and lower bounds like this:

```
parameter_df = petab.parameters.create_parameter_df(petab_problem.sbml_model,
                                                    petab_problem.condition_df,
                                                    petab_problem.observable_df,
                                                    petab_problem.measurement_df) # take parameters from SBML
parameter_df['lowerBound'] = parameter_df['nominalValue'] # set lowerBound to 1e-2*parameter_df
parameter_df['upperBound'] = parameter_df['nominalValue'] # set upperBound to 1e+2*parameter_df
```

You can see that I have moved the upper and lower bounds to equal the initial value, so all parameters should be fixed. Since a time course using these values succeeds in Copasi, I expect it to succeed in AMICI as well. However, I am still getting these warnings:

```
<amici.amici.ParameterScalingVector; proxy of <Swig Object of type 'std::vector< enum amici::ParameterScaling,std::allocator< enum amici::ParameterScaling > > *' at 0x000001C22DCBBEA0> >
  0%|                                                                                                                                                                                               | 0/10 [00:00<?, ?it/s][Warning] AMICI:simulation: AMICI simulation failed:
Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed: No convergence was achieved. Second run of Newton solver failed: No convergence was achieved.
Error occurred in:
stacktrace not available on windows platforms

[Warning] AMICI:simulation: AMICI simulation failed:
Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed: No convergence was achieved. Second run of Newton solver failed: No convergence was achieved.
Error occurred in:
stacktrace not available on windows platforms

[Warning] AMICI:NaN: AMICI encountered a NaN value at index 0/286 in fxdot!
[Warning] AMICI:NaN: AMICI encountered a NaN value at index 0/696 in p!
[Warning] AMICI:NaN: AMICI encountered a NaN value at index 57/515 in w!
[Warning] AMICI:CVODES:CVode:OTHER: AMICI ERROR: in module CVODES in function CVode : The right-hand side routine failed at the first call. 
[Warning] AMICI:simulation: AMICI simulation failed:
Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed. Second run of Newton solver failed: No convergence was achieved.
Error occurred in:
stacktrace not available on windows platforms
```

I also looked at the last suggestions.
amici.py line 306 is a routine named call_unprocessed. That name doesn't tell me much, but I had already set a breakpoint at it, because I found it in the call stack.

Then I looked at pypesto/objective/base.py line 98. That has a name of \_\_call\_\_, which I understand even less. I tried adding a call to
pypesto.objective, but I don't see how to set the parameters:

```
result = optimize.minimize(problem=problem, optimizer=optimizer, n_starts=10, engine=engine)
result.optimize_result.get_for_key('fval')
ref = visualize.create_references(
    x=petab_problem.x_nominal_scaled, fval=obj(petab_problem.x_nominal_scaled))

visualize.waterfall(result, reference=ref, scale_y='lin')
visualize.parameters(result, reference=ref)

#pypesto.objective()
```

In summary, simulation and steady-state calculation succeed in Copasi, and I can see that the Newton solver in AMICI fails twice, which looks normal, because Copasi also needs to run Newton and forward simulation multiple times. But then AMICI suddenly has NaNs everywhere, so this fails much worse. In Copasi, forward simulation and steady-state calculation succeed, but they fail in PETab/pyPESTO/AMICI. Up to now, I have not been able to determine if the problem is in the model, PETab, pyPESTO, or AMICI. petablint runs correctly. This is also issue #1371 on AMICI.

**Comment by yannikschaelte** [comment]
I meant directly calling `problem.objective(SOME_PARAMETER, return_dict=True)`.

**Comment by thomassligon** [comment]
Thanks, I tried adding this line:
```
test = problem.objective(x=petab_problem.x_nominal_scaled, return_dict=True)
```
so now I have
```
problem = importer.create_problem(obj)

#optimizer = pypesto.ScipyOptimizer()
optimizer = optimize.ScipyOptimizer()

# engine = pypesto.SingleCoreEngine()
#engine = pypesto.MultiProcessEngine()
engine = engine.SingleCoreEngine()

# do the optimization
#result = pypesto.minimize(problem=problem, optimizer=optimizer,
#                          n_starts=10, engine=engine)
result = optimize.minimize(problem=problem, optimizer=optimizer, n_starts=10, engine=engine)
result.optimize_result.get_for_key('fval')
ref = visualize.create_references(
    x=petab_problem.x_nominal_scaled, fval=obj(petab_problem.x_nominal_scaled))

visualize.waterfall(result, reference=ref, scale_y='lin')
visualize.parameters(result, reference=ref)

test = problem.objective(x=petab_problem.x_nominal_scaled, return_dict=True)
```

So, one good answer leads to more stupid questions:
Where should this code be?
What should be in the parameter?

**Comment by yannikschaelte** [comment]
Can you run these lines (more or less)?

```python
problem = importer.create_problem(obj)
test = problem.objective(x=petab_problem.x_nominal_scaled, return_dict=True)
print(test['rdatas'][0])
```

**Comment by yannikschaelte** [comment]
See e.g. https://pypesto.readthedocs.io/en/latest/example/petab_import.html

**Comment by thomassligon** [comment]
Thanks! Today, I only had time for a quick test, and it shows:
```
<amici.amici.ParameterScalingVector; proxy of <Swig Object of type 'std::vector< enum amici::ParameterScaling,std::allocator< enum amici::ParameterScaling > > *' at 0x000001E6DF317EA0> >
[Warning] AMICI:CVODES:CVode:OTHER: AMICI ERROR: in module CVODES in function CVode : At t = 3.18384e+199, the right-hand side routine failed in an unrecoverable manner. 
[Warning] AMICI:simulation: AMICI simulation failed:
Steady state computation failed. First run of Newton solver failed: No convergence was achieved. Simulation to steady state failed. Second run of Newton solver failed: No convergence was achieved.
Error occurred in:
stacktrace not available on windows platforms

<amici.numpy.ReturnDataView object at 0x000001E6DF335548>
```

The time looks interesting: t = 3.18384e+199 (seconds) is a very long time. If we need that kind of time, I could use Copasi to simulate to higher and higher time. Up to now, I only tried up to e+9, but I can easily extend this and see if any species grows after such a long time.

**Comment by jvanhoefer** [comment]
Hey Tom, 
As I understand the error message, the problem seems to be the steady state simulation, maybe you have data, that is in the steady state (`t=inf`) or you maybe use pre-equilibration? The first question would be, if your model does contain a steady state, or If it is oscillatory or explodes. Simulating in COPASI for the same parameters could tell you that. Otherwise, as this is an issue regarding the AMICI steady state solver, @paulstapor might have some tipps&tricks available... :)

**Comment by thomassligon** [comment]
Thanks again to both of you! @yannikschaelte , your last suggestion, adding that line of code, is very useful. Some more quick tests in Copasi point to the problem, but finding it will take some time because of the size of the model. Basically, since Copasi steady state was succeeding, and simulation up to 1e+9 seconds looked good (very flat curves), it appeared as though the model was basically working. However, simulating up to 1e+20 seconds uncovers the problem: there are species that grow too much, but so slowly that it isn't visible after 1e+9. Simulation to 1e+20 and 1e+200 show the same behavior. Just for comparison, here is the error I get from Copasi when I simulate to 1e+20:
```
>EXCEPTION 2021-05-18T10:13:38<
  CTrajectoryMethod (6): Deterministic integration failed. LSODA reported:
  DLSODA-  At T (=R1) and step size H (=R2), the    
        corrector convergence failed repeatedly     
        or with ABS(H) = HMIN   
  	In above message, R1 = '1.55664e+19', R2 = '2.43635e+07'
   
  
  Please see result for indications of numerical instability.
```
